### ISSUE TITLE
Clients IPs are not changed with DHCP

### ISSUE THREAD

**Issue by MohsenBs** [noxrepo/pox#282]
Hi

I am using Meininet that install on Ubuntu 20.04 

 - I have run dhcp with POX command (( proto.dhcpd --network=10.1.1.0/24 --ip=10.1.1.1 --first=10 --last=None --router=None --dns=4.2.2.1))

- Hosts can get a dynamic address but the address does not change to a different address when DHCP is requested again

- after dhclient and dhclient -r commands the IP address is still the same on mininet hosts. How can change or renew with different IPs?

![image](https://user-images.githubusercontent.com/117582476/201626548-50ca3fc4-5aec-418a-a5a0-d12d281f4848.png)
![image](https://user-images.githubusercontent.com/117582476/201626734-a6fb6d14-506e-401d-89b1-cc923059b074.png)

**Comment by MurphyMc** [comment]
If I remember right, POX's default address pool implementation rotates the entries it serves by default, so this is probably because dhclient itself is *requesting* the same address and POX is granting it.  You could confirm this by watching the DHCP traffic with Wireshark or something.

It wouldn't be too hard to tweak POX's server to not allow the client to re-request the same address.  But if all you want to do is get a new address, I think in the past I have done this by just messing with dhclient.  Specifically, dhclient keeps track of the addresses it has previously leased in a file somewhere.  You could just delete this file before requesting a new address, but I think I've had success telling it to use /dev/null as the file, since that obviously can't actually save the lease information.

TLDR: I think if you stick `-lf /dev/null` into your dhclient command line, it might do what you want.

You could also try using udhcpc (from busybox) instead of dhclient.  I've found it better suited for experimental configurations.

**Comment by MohsenBs** [comment]
> 
Thank you for your continuous support that gives us motivation to learn.

I use  (-lf /dev/null)  as you see still takes the same address
<img width="284" alt="image" src="https://user-images.githubusercontent.com/117582476/202147603-794ff9f1-d49f-403a-88bd-0e3640d3fe3c.png">

**Comment by MohsenBs** [comment]
> If I remember right, POX's default address pool implementation rotates the entries it serves by default, so this is probably because dhclient itself is _requesting_ the same address and POX is granting it. You could confirm this by watching the DHCP traffic with Wireshark or something.
> 
> It wouldn't be too hard to tweak POX's server to not allow the client to re-request the same address. But if all you want to do is get a new address, I think in the past I have done this by just messing with dhclient. Specifically, dhclient keeps track of the addresses it has previously leased in a file somewhere. You could just delete this file before requesting a new address, but I think I've had success telling it to use /dev/null as the file, since that obviously can't actually save the lease information.
> 
> TLDR: I think if you stick `-lf /dev/null` into your dhclient command line, it might do what you want.
> 
> You could also try using udhcpc (from busybox) instead of dhclient. I've found it better suited for experimental configurations.


I also use the below command :
$ dhclient -r
$ rm /var/lib/dhcp/dhclient*   
$ dhclient

same problem

**Comment by MurphyMc** [comment]
Try editing pox/proto/dhcpd.py line 444.

It's currently:
```
        offer = pool[0]
```
Try this instead:
```
        import random
        offer = random.choice(pool)
```

**Comment by MohsenBs** [comment]
> -lf /dev/null
The address has not changed, but it is taking random, not sequential

<img width="389" alt="image" src="https://user-images.githubusercontent.com/117582476/202563470-b4e41211-a91a-4659-82f2-abe508aea99b.png">
<img width="250" alt="image" src="https://user-images.githubusercontent.com/117582476/202563577-1dd1d312-a65e-4f6e-91ee-27073d82452e.png">

**Comment by MurphyMc** [comment]
POX is not logging any releases.  This is because dhclient is not successfully sending it because it tries to unicast it to the server, which requires that it ARP for the server's Ethernet address.  But there is no actual server to respond to the ARP request.  Run the POX component `proto.arp_responder --no-learn --10.0.0.254` (changing the IP address to be whatever your DHCP server's IP address is supposed to be).

I tested this and it basically worked, *but* it turned up a bug in dhcpd (I don't think I've ever had a reason to actually use releases).  I tried to fix it and pushed it to halosaur.
